Refactored receptor class (chorizo) to generalize beyond proteins

diogomart opened the pull request:
The changes are extensive, and include:
 - every chorizo residue has its own MoleculeSetup instance, whether flexible (sidechain) or not
 - refactored flexibility.py

diogomart committed:
add (failing) test for new residue IDs (icode, no resname)

psharma-sr committed:
removed extra function from LinkedRDKitChorizo, started commenting molsetup

psharma-sr committed:
more comments added to molsetups

psharma-sr committed:
MoleculeSetupEncoder added

psharma-sr committed:
Encoders complete

psharma-sr committed:
added object hooks

psharma-sr committed:
changed molsetup import

psharma-sr committed:
modified decoder names

diogomart committed:
to_json and from_json wrappers, json fo rdkit mols

psharma-sr committed:
midway through writing tests for json_serialization, fixing bugs as I go

psharma-sr committed:
a test written, many bugs fixed, code not clean yet

psharma-sr committed:
some cleanup in molsetups and tests

diogomart committed:
refactor to generalize chorizo beyond protein

psharma-sr committed:
Some small bug fixes and formatting changes, tests failing

diogomart committed:
a couple little bug fixes

sforli committed:
Added support for ProDy parsing of PDB and mmCIF

ProDy functions are defined in utils/prodyutils.py.
The script mk_prepare_receptor.py has been updated with a `--macromol`
option to specify PDB or mmCIF files.

diogomart committed:
minor fixes to --macromol (prody) in mk_prepare_receptor.py

diogomart committed:
minor fixes

diogomart committed:
docstrings for chorizos

diogomart committed:
Merge remote-tracking branch 'origin/develop' into residue_molsetups

diogomart committed:
use correct template with set_template in chorizo

diogomart committed:
docstring for chorizo's ResidueTemplate

diogomart committed:
fix compromised H positions on backbone N

diogomart committed:
docstring for chorizo's update_H_positions()

diogomart committed:
minor notes about chirality of Hs

diogomart committed:
minor fix in chorizo export of static positions

diogomart committed:
DNA templates, mk_prepare_receptor --add_templates

diogomart committed:
atom names in chorizo residues

diogomart committed:
atom names for nucleic acids

diogomart committed:
fix atom names of nucleic acids

diogomart committed:
RNA G 3-end template

diogomart committed:
method to reparameterize existing chorizo

psharma-sr committed:
added DS_Store to gitignore

psharma-sr committed:
updated ChorizoResidue, fixed test LinkedRDKitChorizo call

psharma-sr committed:
ran Black formatting on a few files

psharma-sr committed:
bugfix in json serialization test fixtures

psharma-sr committed:
bugfixing in json serialization tests

psharma-sr committed:
Found json serialization bug, decoder works again

psharma-sr committed:
added test for chorizo residue encoding and decoding
?

psharma-sr committed:
Untested Encoders and Decoders for LinkedRDKitChorizo

diogomart committed:
fix tiny typos in error messages

psharma-sr committed:
Wrote basic test cases for all LinkedRDKitChorizo serialization, fixed bugs

psharma-sr committed:
Black changes to linked_rdkit_chorizo

psharma-sr committed:
Updated comments

psharma-sr committed:
Merge pull request #83 from forlilab/json_serialization

Json Encoders and Decoders in Molsetups and Chorizo Residues

diogomart committed:
refactor flexibility

- ignored atoms are excluded from flexibility model
- ring bonds remain rotatable during macrocycle search
- graph depth considers number of atoms in each rigid group
- broken bond in macrocycle is not deleted and is set as rotatable

diogomart committed:
test flexibility refactor, one fails due to old bug

diogomart committed:
fix tiny errors

diogomart committed:
some metals and ongoing flexres (re-)implementation

diogomart committed:
flexible sidechains

diogomart committed:
fixed most chorizo tests, two still need work

diogomart committed:
move inter residue bonds out of chorizo's init

psharma-sr commented on meeko/__init__.py:
```diff
@@ -47,21 +55,22 @@
+           'pdbutils', 'geomutils', 'rdkitutils', 'utils',
+           'AtomTyper', 'PDBQTMolecule', 'PDBQTReceptor', 'analysis',
+           'LinkedRDKitChorizo', 'ChorizoResidue', 'ResiduePadder', 'ResidueTemplate',
+           'LinkedRDKitChorizoEncoder', 'ChorizoResidueEncoder', 'ResiduePadderEncoder', 'ResidueTemplateEncoder',
```
ResidueChemTemplates and the ResidueChemTemplatesEncoder should probably be added here

psharma-sr commented on meeko/data/residue_chem_templates.json:
```diff
@@ -0,0 +1,567 @@
+{
```
We should remember to include a description of what this data file is in documentation. I'm putting this comment here as a reminder, I'll add that to the documentation issue more explicitly and resolve this comment later.

psharma-sr commented on meeko/flexibility.py:
```diff
@@ -5,275 +5,269 @@
+    molsetup,
+    root_atom_index=None,
+    break_combo_data=None,
+    glue_pseudo_atoms=None,
```
glue_pseudo_atoms is not used anywhere in this function, should it be deleted or should we keep it for future functionality

diogomart replied:
Will remove, it's only needed in `update_closure_atoms` where it's called `coords_dict` (will rename to `glue_pseudo_atoms` for consistency). Just for reference here's a test for this feature: https://github.com/forlilab/Meeko/blob/6317966790b4e9817e37649bf771adb13bc3d921/test/macrocycle_test.py#L35

diogomart replied:
removed in ac2da0724891385abdebe0a207e7a1fb7a202453

psharma-sr commented on meeko/flexibility.py:
```diff
@@ -5,275 +5,269 @@
-        """
-        if len(self.flexibility_models) == 1: # no macrocyle open rings
-            best_model = list(self.flexibility_models.values())[0]
+def _calc_max_weighted_depth(model, seed_node, bonds_to_break, visited=None, depth=0):
```
This file probably needs a few function level comments, but I think I can handle some of that to get it started/make it easier

diogomart replied:
added a couple in ac2da0724891385abdebe0a207e7a1fb7a202453

psharma-sr commented on meeko/flexibility.py:
```diff
@@ -5,275 +5,269 @@
+            and group_index != flex_model["root"]
+        ):
+            remove[group_index] = members[0]
+    flex_model["rigid_body_count"] -= len(remove)
```
Would we want to consider updating the count after the removal has happened? How likely is it that there would be an error during the removal for loop?

diogomart replied:
If there's an error mid loop and it's handled with an except statement, the entire molecule setup would be compromised and should be discarded. In any case, I'll update the count after the for loop in case someone ever thinks of using the count as a proxy for successful completion.

diogomart replied:
added in ac2da0724891385abdebe0a207e7a1fb7a202453

psharma-sr commented on meeko/flexibility.py:
```diff
@@ -5,275 +5,269 @@
+            raise RuntimeError(msg)
+        root_body_index = get_root_body_index(flex_model, root_atom_index)
+        flex_model["root"] = root_body_index
+        broken_bonds = []
```
Is there a reason we're bothering to return broken_bonds if it's an empty list always anyway?

diogomart replied:
that's the first return statement of this function, the second one returns non-empty lists: https://github.com/forlilab/Meeko/blob/6317966790b4e9817e37649bf771adb13bc3d921/meeko/flexibility.py#L125-L126

diogomart committed:
some linting on flexibility.py

psharma-sr commented on meeko/flexibility.py:
```diff
@@ -5,275 +5,279 @@
+            [not molsetup.atom_ignore[i] for i in range(len(molsetup.atom_ignore))]
+        )
+        if len(model["visited"]) != nr_not_ignored:
+            msg = f"{len(model['visited'])=} differs from not-ignored atoms {nr_not_ignored}"
```
Is this runtime error not being exactly like the first intended? Do we still need the print statement in the earlier RuntimeError?

diogomart replied:
That's right, the errors are identical. Removed print statements in 41247d2866a5b0a3bab4825685aeddbb1289d4a0

psharma-sr commented on meeko/flexibility.py:
```diff
@@ -5,275 +5,279 @@
+                unbroken_rings_bonds.append(bond)
+
+        model = walk_rigid_body_graph(molsetup, bond_break_combo, unbroken_rings_bonds)
+        nr_not_ignored = sum(
```
I don't think this would be changing with the loop, but maybe I'm missing something. Could it just be instantiated outside the loop?

diogomart committed:
more linting on flexibility.py

diogomart committed:
bonds between residues used for matching templates

 - all tests passing now
 - blunt ends are required explicitly, this can be annoying

diogomart committed:
icode and linting on chorizo

psharma-sr committed:
black formatting, added more detailed comments and space for comments in many files and made comment formatting more consistent across files, cleaned up simple syntax errors

psharma-sr commented on meeko/linked_rdkit_chorizo.py:
```diff
@@ -226,291 +517,783 @@ class LinkedRDKitChorizo:
+        # currently allowing only one bond per residue pair
+        if any([len(v) > 1 for k, v in bonds.items()]):
+            msg = "got more than one bond between some residue pairs:" + os_linesep
+            for key, value in repeated.items():
```
I don't think repeated is defined here, not sure what it should be though

psharma-sr commented on meeko/macrocycle.py:
```diff
@@ -231,15 +364,18 @@ def show_macrocycle_scores(self, setup):
             bond_by_ring = defaultdict(list)
```
It doesn't look like this function is being used anywhere, so I have deprecated it for now. All this logic here is after a return statement and is therefore inaccessible code, so if we did want to keep this function we should give that some thought.

psharma-sr committed:
Added ResidueChemTemplates and Encoder to the init

diogomart committed:
minor fix in error message

diogomart committed:
add a couple missing RNA templates

diogomart committed:
add ACE and NME templates

diogomart committed:
chorizo json serialization with None rdkit mols

diogomart committed:
fixes to JSON (de)serialization

diogomart committed:
bug fix in chorizo export_static_atom_params
